### PR TITLE
Generate Scaffolding for Bindings Features

### PR DESCRIPTION
**Pull request by hjkatz**
closes: https://github.com/ngrok-private/ngrok/issues/32542

## What

Add the scaffolding for new bindings features

See: https://github.com/ngrok-private/ngrok/issues/32542

## How

Use `kubebuilder` CLI

Some example output
```
$ kb describe operatorconfiguration ngrok-operator
Name:         ngrok-operator
Namespace:    ngrok-operator
Labels:       <none>
Annotations:  <none>
API Version:  ngrok.k8s.ngrok.com/v1beta1
Kind:         OperatorConfiguration
Metadata:
  Creation Timestamp:  2024-09-16T20:18:05Z
  Generation:          2
  Resource Version:    123877
  UID:                 b3bf2aec-dc16-4341-90bf-4dbf476332f3
Spec:
  API URL:         https://api.ngrok.com
  App Version:     0.12.2
  Cluster Domain:  svc.cluster.local
  Description:     The official ngrok Kubernetes Operator.
  Enabled Features:
    ingress
    bindings
  Id:        k8_example123
  Metadata:  env=local,from=makefile
  Uri:       https://api.ngrok.com/kubernetes_operators/k8_example123
Events:      <none>
```

```
$ kb get operatorconfigurations.ngrok.k8s.ngrok.com ngrok-operator
NAME             APP VERSION   ID              DESCRIPTION
ngrok-operator   0.12.2        k8_example123   The official ngrok Kubernetes Operator.
```

Startup logs:
```
...
2024-09-16T21:03:15Z	INFO	setup	OperatorConfiguration created	{"result": "unchanged"}
2024-09-16T21:03:15Z	INFO	setup	Endpoint Bindings controller enabled
2024-09-16T21:03:15Z	INFO	setup	starting manager
...
```

## Breaking Changes
*Are there any breaking changes in this PR?*

### CODE REVIEW

File: internal/controller/bindings/bindingconfiguration_controller.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *BindingConfigurationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&bindingsv1alpha1.BindingConfiguration{}).
+		Complete(r)
+}
```
Comment (hjkatz) [inline internal/controller/bindings/bindingconfiguration_controller.go:77]: Does this need some owned by things or anything like that?

File: api/bindings/v1alpha1/bindingconfiguration_types.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	// Support more patterns in the future, see product spec
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:items:Pattern=`^[*]$`
+	AllowedURLs []string `json:"allowed_urls"`
+
+	// CertsSecretName is the name of the k8s secret that contains the TLS certificate, CSR, private/public keys to use for the ngrok API
+	// TODO(hkatz) Create controller to manage this Secret lifecycle
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="global-binding-configuration-certs"
+	CertsSecretName string `json:"certs_secret_name"`
```
Comment (hjkatz) [inline api/bindings/v1alpha1/bindingconfiguration_types.go:69]: Instead of storing the `PublicKey` here directly I chose to use a Secret reference. Is this how we do the references? Or is there another pattern I should use?

File: internal/controller/bindings/bindingconfiguration_controller.go (hunk 1 above)
Comment (jonstacks) [reply internal/controller/bindings/bindingconfiguration_controller.go:77]: If this will be creating and managing our endpoint bindings, then this will probably need to own those

File: api/bindings/v1alpha1/bindingconfiguration_types.go (hunk 2 above)
Comment (jonstacks) [reply api/bindings/v1alpha1/bindingconfiguration_types.go]: We currently have this pattern if you are extracting a specific key out of a k8s secret: https://github.com/ngrok/ngrok-operator/blob/1e2367de30c22a043fac83b7c946eb24925d4423/api/ingress/v1alpha1/ngrok_common.go#L217

It sounds like this reference is for creating a new k8s secret with `type=kubernetes.io/tls`. Is that correct?

File: internal/controller/bindings/bindingconfiguration_controller.go (hunk 1 above)
Comment (hjkatz) [reply internal/controller/bindings/bindingconfiguration_controller.go:77]: Yes, then it will. I'll take a look how to do that, maybe pairing on it would be quickest though.

File: api/bindings/v1alpha1/bindingconfiguration_types.go (hunk 2 above)
Comment (hjkatz) [reply api/bindings/v1alpha1/bindingconfiguration_types.go]: I'm not sure if it's exactly `type=kubernetes.io/tls` because we want to store the CSR too, but I do think I can use this `ClientSecret *SecretKeyRef` pattern for this field. Thanks!

**Comment by hjkatz** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-operator/427?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#427** <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-operator/427?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#430** <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-operator/430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hjkatz and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: cmd/main.go [hunk 3]
```diff
@@ -116,10 +126,12 @@ func cmd() *cobra.Command {
 		},
 	}
 
+	c.Flags().StringVar(&opts.appVersion, "app-version", "0.0.0", "AppVersion provided by the Helm Chart")
```
Comment (jonstacks) [inline cmd/main.go:129]: I think we can remove this and just pull this from the version package https://github.com/ngrok/ngrok-operator/blob/e1fbfc88940eea5b317d7e880100bea00369198b/internal/version/version.go#L29-L31

File: helm/ngrok-operator/README.md [hunk 4]
```diff
@@ -59,6 +60,7 @@ To uninstall the chart:
 | `ingressClass.create`                | Whether to create the ingress class.                                                                                  | `true`                             |
 | `ingressClass.default`               | Whether to set the ingress class as default.                                                                          | `false`                            |
 | `controllerName`                     | The name of the controller to look for matching ingress classes                                                       | `k8s.ngrok.com/ingress-controller` |
+| `useExperimentalGatewayApi`          | When true, enbale the Gateway controller                                                                              | `false`                            |
```
Comment (jonstacks) [inline helm/ngrok-operator/README.md:63]: **Non-blocking**: I've been thinking about this more as we add more features to the operator. I think we could take a page out of the datadog operator's playbook and nest these all under a `features` key

File: helm/ngrok-operator/templates/controller-deployment.yaml [hunk 5]
```diff
@@ -61,6 +61,10 @@ spec:
         command:
         - /manager
         args:
+        - --app-version={{ .Chart.AppVersion }}
```
Comment (jonstacks) [inline helm/ngrok-operator/templates/controller-deployment.yaml:64]: Left a note on this up above. I think we can remove this. This app version is just the default for the helm chart and it can always be changed by supplying a different image.

File: api/bindings/v1alpha1/endpointbinding_types.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+// NOTE: Run "make" to regenerate code after modifying this file
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// EndpointBindingSpec defines the desired state of EndpointBinding
+type EndpointBindingSpec struct {
+	// TargetService is the name of the Service that this Endpoint projects
+	// +kubebuilder:validation:Required
+	TargetService string `json:"targetService"`
```
[the remark this answers is not in the document]
Comment (hjkatz) [reply api/bindings/v1alpha1/endpointbinding_types.go:38]: Sure!

File: cmd/main.go (hunk 3 above)
Comment (hjkatz) [reply cmd/main.go:129]: Does this match `.Chart.AppVersion`? I'm seeing `VERSION` in the root of the repo, but how is that updated? Is that the Release version or the App version?

File: helm/ngrok-operator/README.md (hunk 4 above)
Comment (hjkatz) [reply helm/ngrok-operator/README.md:63]: I have a ticket for this here: https://github.com/ngrok-private/ngrok/issues/32951

tl;dr:
```
# values.yaml
ingress:
  enabled: true/false

gateway:
  enabled: true/false

bindings:
  enabled: true/false
```

File: cmd/main.go (hunk 3 above)
Comment (jonstacks) [reply cmd/main.go]: The `VERSION` file in the root of the repo is used to trigger the docker image release workflow. We also use that file as part of the ldflags to the version package. This makes sure that the binary in the docker image has the version information. While the chart's AppVersion is used as the default image tag if one is not specified, user's can overwrite this and pull an older or newer image, so just passing it through might be wrong in certain cases.

**Review by jonstacks — APPROVED** [review]

**Comment by hjkatz** [comment]
### Merge activity

* **Sep 17, 3:11 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/ngrok/ngrok-operator/427) rebased this pull request as part of a merge.
